src/black/trans.py

```diff
@@ -371,30 +376,50 @@ def do_match(self, line: Line) -> TMatchResult:
 
         is_valid_index = is_valid_index_factory(LL)
 
-        for i, leaf in enumerate(LL):
+        string_indices = []
+        idx = 0
+        while is_valid_index(idx):
```

Comment:
Probably best for another PR, but there may be an optimization opportunity here. `is_valid_index(idx)` is basically equivalent to `idx < len(LL)`, but it does so through some indirection and a nested function. According to Jukka mypyc isn't good at compiling nested functions, so this code may get a good speedup if we just use `idx < len(LL)`.

Reply:
I tried this (https://github.com/JelleZijlstra/black/commit/9b1f0b11ffd449d6cbf9f2326e316257a707a008) but didn't see a significant difference under ESP, though I didn't do rigorous benchmarking.